pkg/registry/clusteroperator/clusterversion/strategy.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterversion
+
+import (
+	"github.com/openshift/cluster-operator/pkg/api"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/apiserver/pkg/storage/names"
+
+	"github.com/golang/glog"
+	"github.com/openshift/cluster-operator/pkg/apis/clusteroperator"
+	"github.com/openshift/cluster-operator/pkg/apis/clusteroperator/validation"
+)
+
+// NewScopeStrategy returns a new NamespaceScopedStrategy for clusters
+func NewScopeStrategy() rest.NamespaceScopedStrategy {
+	return clusterRESTStrategies
+}
+
+// implements interfaces RESTCreateStrategy, RESTUpdateStrategy, RESTDeleteStrategy,
+// NamespaceScopedStrategy
+type clusterRESTStrategy struct {
+	runtime.ObjectTyper // inherit ObjectKinds method
+	names.NameGenerator // GenerateName method for CreateStrategy
+}
+
+// implements interface RESTUpdateStrategy
+type clusterVersionStatusRESTStrategy struct {
+	clusterRESTStrategy
+}
+
+var (
+	clusterRESTStrategies = clusterRESTStrategy{
+		// embeds to pull in existing code behavior from upstream
+
+		// this has an interesting NOTE on it. Not sure if it applies to us.
+		ObjectTyper: api.Scheme,
+		// use the generator from upstream k8s, or implement method
+		// `GenerateName(base string) string`
+		NameGenerator: names.SimpleNameGenerator,
+	}
+	_ rest.RESTCreateStrategy = clusterRESTStrategies
+	_ rest.RESTUpdateStrategy = clusterRESTStrategies
+	_ rest.RESTDeleteStrategy = clusterRESTStrategies
+
+	clusterVersionStatusUpdateStrategy = clusterVersionStatusRESTStrategy{
+		clusterRESTStrategies,
+	}
+	_ rest.RESTUpdateStrategy = clusterVersionStatusUpdateStrategy
+)
+
+// Canonicalize does not transform a cluster.
+func (clusterRESTStrategy) Canonicalize(obj runtime.Object) {
+	_, ok := obj.(*clusteroperator.ClusterVersion)
+	if !ok {
+		glog.Fatal("received a non-cluster version object to create")
+	}
+}
+
+// NamespaceScoped returns true as cluster versions are scoped to a namespace.
+func (clusterRESTStrategy) NamespaceScoped() bool {
+	return true
+}
+
+func (clusterRESTStrategy) PrepareForCreate(ctx genericapirequest.Context, obj runtime.Object) {
```

Comment:
Need to clear out the status. A user cannot set the status when creating an object.